llama-index-core/llama_index/core/agent/runner/planner.py

```diff
@@ -0,0 +1,544 @@
+        return next_sub_tasks
+
+    def get_remaining_subtasks(self, plan_id: str) -> List[SubTask]:
+        remaining_subtasks = []
+        plan = self.plan_dict[plan_id]
+
+        if plan_id not in self.completed_sub_tasks:
+            self.completed_sub_tasks[plan_id] = []
+
+        completed_sub_tasks = self.completed_sub_tasks[plan_id]
+        completed_sub_task_names = [sub_task.name for sub_task in completed_sub_tasks]
+
+        for sub_task in plan.sub_tasks:
+            if sub_task.name not in completed_sub_task_names:
+                remaining_subtasks.append(sub_task)
+        return remaining_subtasks
+
+    def reset(self) -> None:
+        """Reset."""
+        self.task_dict = {}
+        self.completed_sub_tasks = {}
+        self.plan_dict = {}
+
+
+DEFAULT_INITIAL_PLAN_PROMPT = """\
+Think step-by-step. Given a task and a set of tools, create a comprehesive, end-to-end plan to accomplish the task.
+Keep in mind not every task needs to be decomposed into multiple sub-tasks if it is simple enough.
+The plan should end with a sub-task that satisfies the overall task.
+
+The tools available are:
+{tools_str}
+
+Overall Task: {task}
+"""
+
+DEFAULT_PLAN_REFINE_PROMPT = """\
+Think step-by-step. Given an overall task, a set of tools, and completed sub-tasks, update (if needed) the remaining sub-tasks so that the overall task can still be completed.
+The plan should end with a sub-task that satisfies the overall task.
+If the remaining sub-tasks are sufficient, you can skip this step.
+
+The tools available are:
+{tools_str}
+
+Overall Task:
+{task}
+
+Completed Sub-Tasks + Outputs:
+{completed_outputs}
+
+Remaining Sub-Tasks:
+{remaining_sub_tasks}
+"""
+
+
+class StructuredPlannerAgent(AgentRunner):
+    """Structured Planner Agent runner.
+
+    Top-level agent orchestrator that can create tasks, run each step in a task,
+    or run a task e2e. Stores state and keeps track of tasks.
+
+    Args:
+        agent_worker (BaseAgentWorker): step executor
+        chat_history (Optional[List[ChatMessage]], optional): chat history. Defaults to None.
+        state (Optional[AgentState], optional): agent state. Defaults to None.
+        memory (Optional[BaseMemory], optional): memory. Defaults to None.
+        llm (Optional[LLM], optional): LLM. Defaults to None.
+        callback_manager (Optional[CallbackManager], optional): callback manager. Defaults to None.
+        init_task_state_kwargs (Optional[dict], optional): init task state kwargs. Defaults to None.
+
+    """
+
+    def __init__(
+        self,
+        agent_worker: BaseAgentWorker,
+        tools: Optional[List[BaseTool]] = None,
+        tool_retriever: Optional[ObjectRetriever[BaseTool]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        state: Optional[PlannerAgentState] = None,
+        memory: Optional[BaseMemory] = None,
+        llm: Optional[LLM] = None,
+        initial_plan_prompt: str = DEFAULT_INITIAL_PLAN_PROMPT,
```

Comment:
nit: most of our existing fn signatures take in Optional[PromptTemplate] as input. If you still want to allow strings could do Optional[Union[PromptTemplate, str]]

Reply:
Done!